contrib/log_correlation/log4j/src/main/java/io/opencensus/contrib/logcorrelation/log4j/OpenCensusTraceContextDataInjector.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.contrib.logcorrelation.log4j;
+
+import io.opencensus.common.ExperimentalApi;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.core.ContextDataInjector;
+import org.apache.logging.log4j.core.Layout;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.Property;
+import org.apache.logging.log4j.core.impl.ThreadContextDataInjector;
+import org.apache.logging.log4j.util.SortedArrayStringMap;
+import org.apache.logging.log4j.util.StringMap;
+
+/**
+ * A Log4j {@link ContextDataInjector} that adds OpenCensus tracing data to log events.
+ *
+ * <p>This class adds the following key-value pairs:
+ *
+ * <ul>
+ *   <li>{@value #TRACE_ID_CONTEXT_KEY} - the lowercase base16 encoding of the current trace ID
+ *   <li>{@value #SPAN_ID_CONTEXT_KEY} - the lowercase base16 encoding of the current span ID
+ *   <li>{@value #TRACE_SAMPLED_CONTEXT_KEY} - the sampling decision of the current span ({@code
+ *       "true"} or {@code "false"})
+ * </ul>
+ *
+ * <p>The tracing data can be accessed with {@link LogEvent#getContextData} or included in a {@link
+ * Layout}. For example, the following patterns could be used to include the tracing data with a <a
+ * href="https://logging.apache.org/log4j/2.x/manual/layouts.html#Pattern_Layout">Pattern
+ * Layout</a>:
+ *
+ * <ul>
+ *   <li><code>%X{openCensusTraceId}</code>
```

Comment:
PS you can drop-in with some existing stuff by re-using "traceId" and "spanId" at the risk of clash. For example, I don't know how exactly it would work if someone wanted brave and census with same logging config where brave uses "traceId" and here "openCensusTraceId" yet both have literally the same value. That said, many libraries use different names ex sleuth uses the same config keys as B3 header names for legacy reasons (eventhough people can and do override that)

Reply:
Thanks for the suggestion.  I was wondering which key to use, but I thought that it would be better to use longer names and avoid conflicts, since users will probably need to change the field names anyway.  Then they can choose which tracing library's field to remap.  For example, I needed to transform the trace ID field to `logging.googleapis.com/trace` to export to Stackdriver.  What are the main downsides of having multiple field names?  I could think of a few reasons:

- Shorter, more standard names could be clearer.
- Deduplicating tracing fields removes the possibility of inconsistent values.
- There would be fewer entries in the LogEvent, which could help performance.

Another reason that I prefer the longer names is that this is currently marked as an experimental feature, so I don't want it to interfere with other libraries.  Do you think it makes sense to release it with the "openCensus" field names and then agree on final key names in opencensus-specs (https://github.com/census-instrumentation/opencensus-specs/issues/123)?

Reply:
I'm just calling attention to this concern as it rose. I think if you have some cycles to see how someone can use multiple (like fallback this if not present that), or configure the key names, this would be helpful for those migrating or interoping even if not strictly our duty.

Reply:
Now I understand.  I think that allowing users to specify the keys would be useful.  I would rather add it in a follow-up PR, though.

Reply:
I opened an issue: #1389